### PR TITLE
Turn on MessageFormat BlobRecord Version2

### PR DESCRIPTION
**Pull request by nsivabalan**
This patch turns on MessageFormat BlobRecord Version2. So, any new PUTs will be written in the new format. 

Note: Support to read blobs in new format is already in PROD (supports both versions of blobs)

Fixes going in with this patch:
1. Default Put is done using new format (Blob Record Version 2)
2. PutMessageFormatBlobV1InputStream, a test class to put in older format. (this was changed from version 2 to version 1)
3. Changes in MockBlockingChannel and MockDataNode since getBlob() returns BlobData instead of BlobOutput. (BlobData has BlobType in addition to what BlobOutput had)
4. Changes in some tests to adhere to the new format and some which had hard coded log end offsets for testing purposes.

Testing:
./gradlew build && ./gradlew test passed 
Deployed server and frontend locally. Put, get and delete tested. DumpData tested.

Reviewers: 
Gopal, Priyesh
SLA: 15 mins

### CODE REVIEW

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java [hunk 1]
```diff
@@ -106,9 +107,12 @@ public void send(Send send)
         BlobId blobId = putRequest.getBlobId();
         BlobProperties blobProperties = putRequest.getBlobProperties();
         ByteBuffer userMetadata = putRequest.getUsermetadata();
-        BlobOutput blobOutput = new BlobOutput(putRequest.getBlobSize(), putRequest.getBlobStream());
+        // Since this is a test class, we assume that the Inputstream sent in via PutRequest is ByteBufferInputStream
```
Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: Why care about the extra copy in the tests and add this comment? Just create a `ByteBufferInputStream` and keep things simple?

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java [hunk 2]
```diff
@@ -35,36 +32,16 @@
 public class MockDataNode {
   private DataNodeId dataNodeId;
 
-  private Map<BlobId, MaterializedBlob> blobs;
```
Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: why make this change?

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java (hunk 1 above)
Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: ok fine. will fix it

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java (hunk 2 above)
Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: Reason is that, already BlobData has ByteBufferInputStream and hence we don't need an extra processing (since earlier we had to convert InputStream to ByteBufferInputStream). But this raises a question why BlobData has ByteBufferInputStream whereas BlobOutput had InputStream, for which I don't have an answer. Looks like you made the change for it (https://github.com/linkedin/ambry/pull/320/files). Do you know why this change was made. Sorry, I didn't realize.

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 3]
```diff
@@ -210,7 +210,8 @@ protected static void endToEndTest(Port targetPort, String coordinatorDatacenter
       channel.send(getRequestExpired);
       InputStream streamExpired = channel.receive().getInputStream();
       GetResponse respExpired = GetResponse.readFrom(new DataInputStream(streamExpired), clusterMap);
-      Assert.assertEquals(ServerErrorCode.Blob_Expired, respExpired.getPartitionResponseInfoList().get(0).getErrorCode());
+      Assert
+          .assertEquals(ServerErrorCode.Blob_Expired, respExpired.getPartitionResponseInfoList().get(0).getErrorCode());
```
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: is this formatted correctly?

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java [hunk 4]
```diff
@@ -475,7 +475,7 @@ public void endToEndTestHardDeletes()
     notificationSystem.awaitBlobDeletions(blobIdList.get(6).getID());
 
     time.currentMilliseconds = time.currentMilliseconds + Time.SecsPerDay * Time.MsPerSec;
-    ensureCleanupTokenCatchesUp(partitionIds.get(0).getReplicaIds().get(0).getReplicaPath(), mockClusterMap, 297905);
+    ensureCleanupTokenCatchesUp(partitionIds.get(0).getReplicaIds().get(0).getReplicaPath(), mockClusterMap, 297923);
```
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java]: did you run the test in a loop on both Linux and on OS X?

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 5]
```diff
@@ -982,11 +983,11 @@ protected static void endToEndReplicationWithMultiNodeSinglePartitionTest(String
...
       // The token offset value of 13062 was derived as followed:
       // - Up to this point we have done 6 puts and 1 delete
-      // - Each put takes up 2177 bytes in the log (1000 data, 1000 user metadata, 177 ambry metadata)
+      // - Each put takes up 2179 bytes in the log (1000 data, 1000 user metadata, 179 ambry metadata)
       // - Each delete takes up 97 bytes in the log
       // - The offset stored in the token will be the position of the last entry in the log (the delete, in this case)
-      // - Thus, it will be at the end of the 6 puts: 6 * 2177 = 13062
-      checkReplicaTokens(clusterMap, dataNodeId, 13062, "0");
+      // - Thus, it will be at the end of the 6 puts: 6 * 2179 = 13074
+      checkReplicaTokens(clusterMap, dataNodeId, 13074, "0");
```
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: Again, if you haven't, please run this in a loop on Linux and OS X, just to make sure.

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: The description of that PR says why that change was made.
Let us not complicate matters. You shouldn't have to make any change to `MockDataNode.java` or `Blob.java`. I think you can just hardcode `BlobType.DataBlob` in `MockBlockingChannel.java` as I mentioned in the other comment.

**Comment by pnarayanan** [comment]
Finished a pass.

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java [hunk 6]
```diff
@@ -152,13 +156,14 @@ public void send(Send send)
...
+            MockDataNode.BlobDataAndError bdae = mockDataNode.getData(blobId);
             getResponseErrorCode = bdae.getError();
             if (getResponseErrorCode == ServerErrorCode.No_Error) {
-              BlobOutput blobData = bdae.getBlobOutput();
-              byteBufferSize = (int) MessageFormatRecord.Blob_Format_V1.getBlobRecordSize(blobData.getSize());
+              BlobData blobData = bdae.getBlobData();
+              byteBufferSize = (int) MessageFormatRecord.Blob_Format_V2.getBlobRecordSize(blobData.getSize());
               byteBuffer = ByteBuffer.allocate(byteBufferSize);
-              MessageFormatRecord.Blob_Format_V1.serializePartialBlobRecord(byteBuffer, blobData.getSize());
+              MessageFormatRecord.Blob_Format_V2
```
Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: Nope. We need to know the BlobType. So BlobData is noting but (all contents of BlobLOutput) along with BlobType. So ideally it should be size, InputStream and BlobType.

Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: Why do you need to know the `BlobType`? `BlobType` is always `DataBlob` for the coordinator. You can simply construct a `BlobData` from `BlobOutput` and keep things simple.

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java (hunk 2 above)
Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: Nope. I am not complicating things. I went over BlobData and got a clarity that MessageFormat always returns ByteBufferInputStream and hence you have changed the arg in the constructor.

Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: Now let me explain why we don't need materializedBlob. Earlier, when we had BlobOutput(which had just the InputStream), we need materialized Stream which is nothing but a ByteBuffer of the inputStream. So when getBlob() call is made we return it as ByteBufferInputStream from the materialized stream. With this change, we already have ByteBufferInputStream with us. Why would someone convert from convert from A to B, and then again from B to A. A stands for ByteBufferInputStream and B stands for ByteBuffer. A to B happens during PUT and B to A happens during GET. Hope this clarifies things. If not, we can sync up offline.

Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: Siva, for the change in this PR, you should not have to worry about `ByteBufferInputStream` and things like that. This is what I mean by complicating. My claim is that `MockDataNode` will need zero changes.

One thing to keep in mind is that we don't want to make any major changes to the coordinator package - only the minimal amount of changes as required - which means anything that can be avoided should be avoided. We do not want to optimize code that will soon be removed.

Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: Yeah, I can get your point. But I feel that this is not a sweeping or a very big change. Very trivial fix in my opinion.

Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockDataNode.java]: This is an unnecessary change in my opinion.

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java (hunk 6 above)
Comment (nsivabalan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: Somehow I feel that having BlobData would be more intuitive rather than having BlobOutput. Else, we have to add comments everywhere stating that this is legacy code and used only for simple blobs(or old style blobs) and hard to maintain. Making it BlobData makes is more intuitive and don't need of any comments as getBlob() in general returns BlobData

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java (hunk 3 above)
Comment (nsivabalan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: Yes, I did format it.

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java (hunk 4 above)
Comment (nsivabalan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java]: Will check it and let you know.

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java (hunk 6 above)
Comment (pnarayanan) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: Production code for puts in the coordinator is hardcoded today to use `DataBlob`. 

Production code deals with `BlobOutput` today.

Where do you think we have to add comments about legacy code and all? This is a test class. Let us not exaggerate its importance. Coordinator will never deal with metadata blobs in production, why go out of our way to have the tests deal with them?

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java (hunk 4 above)
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java]: You could use the following command to run the server tests ten times:
```
count=0; while (./gradlew :ambry-server:clean && ./gradlew :ambry-server:test && [ $count -ne 10 ]); do ((count++)); printf "\n\nDone with %d iterations\n\n" $count; done
```

Comment (nsivabalan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerHardDeleteTest.java]: yes, ran it in a loop and it passed in both mac and linux

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java (hunk 5 above)
Comment (nsivabalan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: done, passed

File: ambry-messageformat/src/test/java/com.github.ambry.messageformat/MessageFormatInputStreamTest.java [hunk 7]
```diff
@@ -59,9 +59,9 @@ private void messageFormatBlobPropertiesTest(short blobVersion, BlobType blobTyp
     ByteBufferInputStream stream = new ByteBufferInputStream(ByteBuffer.wrap(data));
 
     MessageFormatInputStream messageFormatStream =
-        (blobVersion == MessageFormatRecord.Blob_Version_V1) ? new PutMessageFormatInputStream(key, prop,
+        (blobVersion == MessageFormatRecord.Blob_Version_V2) ? new PutMessageFormatInputStream(key, prop,
```
Comment (pnarayanan) [inline ambry-messageformat/src/test/java/com.github.ambry.messageformat/MessageFormatInputStreamTest.java]: minor: although the check is the same, this and subsequent ones use the ternary condition, whereas the one in `BlobStoreHardDeleteTest.java` uses regular if-else. Perhaps you could make all these checks into switch-case statements?

Comment (nsivabalan) [inline ambry-messageformat/src/test/java/com.github.ambry.messageformat/MessageFormatInputStreamTest.java]: All cases in MessageFormatInputStreamTest are just an assignment statement. do you think we need a switch case for it. I felt ternary is good enough. I can change it for BlobStoreHardDeleteTest which I felt as reasonable

**Comment by pnarayanan** [comment]
You should still use `Blob_Format_V2` everywhere in `MockBlockingChannel` (you seem to have removed the change) as that is the format for any put by the coordinator once this change is enabled. You should hardcode `DataBlob` though as the prod code only writes in that. I am not sure how the tests are passing, but I have not looked deeper into what the tests do.

Basically in `MockBlockingChannel.java` you will need to replace the following:
```java
MessageFormatRecord.Blob_Format_V1.serializePartialBlobRecord(byteBuffer, blobData.getSize());
```
to the following:
```java
MessageFormatRecord.Blob_Format_V2.serializePartialBlobRecord(byteBuffer, blobData.getSize(), BlobType.DataBlob);
```

And while at it, you could perhaps rename the variabl `blobData` to `blobOutput` as its type is `BlobOutput`.

**Comment by nsivabalan** [comment]
@pnarayanan : addressed your comments

File: ambry-messageformat/src/test/java/com.github.ambry.messageformat/BlobStoreHardDeleteTest.java [hunk 8]
```diff
@@ -138,12 +138,15 @@
     private MessageFormatInputStream getPutMessage(StoreKey key, BlobProperties blobProperties, byte[] usermetadata,
         byte[] blob, int blobSize, short blobVersion, BlobType blobType)
         throws MessageFormatException {
-      if (blobVersion == MessageFormatRecord.Blob_Version_V1) {
-        return new PutMessageFormatInputStream(key, blobProperties, ByteBuffer.wrap(usermetadata),
-            new ByteBufferInputStream(ByteBuffer.wrap(blob)), blobSize, blobType);
-      } else {
-        return new PutMessageFormatBlobV2InputStream(key, blobProperties, ByteBuffer.wrap(usermetadata),
-            new ByteBufferInputStream(ByteBuffer.wrap(blob)), blobSize, blobType);
+      switch (blobVersion) {
```
Comment (pnarayanan) [inline ambry-messageformat/src/test/java/com.github.ambry.messageformat/BlobStoreHardDeleteTest.java]: You could make this consistent with the rest of the tests. `switch-case` preferably, but up to you.

**Comment by pnarayanan** [comment]
One comment. LGTM otherwise. @vgkholla please review when you get a chance.

File: ambry-messageformat/src/test/java/com.github.ambry.messageformat/BlobStoreHardDeleteTest.java (hunk 8 above)
Comment (nsivabalan) [inline ambry-messageformat/src/test/java/com.github.ambry.messageformat/BlobStoreHardDeleteTest.java]: Sorry, I dont get you. Based on your suggestion I changed this to a switch case. For other such instances, as I commented earlier, I dint feel switch case is a good fit as its just one assignment statement.

Comment (pnarayanan) [inline ambry-messageformat/src/test/java/com.github.ambry.messageformat/BlobStoreHardDeleteTest.java]: This is no different from the other cases. Not a big deal, but let us just be consistent and use the ternary operator everywhere if that is what you prefer.

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 9]
```diff
@@ -982,11 +983,11 @@ protected static void endToEndReplicationWithMultiNodeSinglePartitionTest(String
       // read the replica file and check correctness
       // The token offset value of 13062 was derived as followed:
```
Comment (cgtz) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: Change this to 13074 as well

File: ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java [hunk 10]
```diff
@@ -155,13 +156,14 @@ public void send(Send send)
             MockDataNode.BlobOutputAndError bdae = mockDataNode.getData(blobId);
             getResponseErrorCode = bdae.getError();
             if (getResponseErrorCode == ServerErrorCode.No_Error) {
-              BlobOutput blobData = bdae.getBlobOutput();
-              byteBufferSize = (int) MessageFormatRecord.Blob_Format_V1.getBlobRecordSize(blobData.getSize());
+              BlobOutput blobOutput = bdae.getBlobOutput();
```
Comment (vgkholla) [inline ambry-coordinator/src/test/java/com.github.ambry.coordinator/MockBlockingChannel.java]: very minor: this should be `boae` keeping it line with the naming convention.

**Comment by nsivabalan** [comment]
@vgkholla : addressed your comments.

**Comment by vgkholla** [comment]
LGTM. Did you do any manual tests? Like bringing up server, making a put request, checking the log to see that it is written in the right format and then doing a get (especially with the coordinator)?

**Comment by nsivabalan** [comment]
Yes, I did some puts, gets and deletes. Also did a data dump to verify that I was able to read the blobs.
